### PR TITLE
feat(profiling): new memory profiler

### PR DESCRIPTION
**Pull request by jd**
## feat(profiling): introduce memalloc collector

This is a new memory allocation profiler that directly leverages the CPython
memory API. It tracks memory allocation events.

## feat(profiling): allow to use memalloc collector

### CODE REVIEW

File: ddtrace/profiling/collector/memalloc.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            super(MemoryCollector, self).stop()
+
+    def collect(self):
+        events, count, alloc_count = _memalloc.iter_events()
+        capture_pct = 100 * count / alloc_count
+        # TODO: The event timestamp is slightly off since it's going to be the time we copy the data from the
+        # _memalloc buffer to our Recorder. This is fine for now, but we might want to store the nanoseconds
+        # timestamp in C and then return it via iter_events.
+        return (
+            tuple(
```
Comment (majorgreys) [inline ddtrace/profiling/collector/memalloc.py:64]: Any particular reason here to use `tuple`?

File: ddtrace/profiling/exporter/pprof.py [hunk 2]
```diff
@@ -138,6 +139,21 @@ def convert_stack_event(self, thread_id, thread_native_id, thread_name, trace_id
...
         self._location_values[location_key]["wall-time"] = sum(s.wall_time_ns for s in samples)
 
+    def convert_memalloc_event(self, thread_id, thread_native_id, thread_name, frames, nframes, events):
+        location_key = (
+            self._to_locations(frames, nframes),
+            (("thread id", str(thread_id)), ("thread native id", str(thread_native_id)), ("thread name", thread_name),),
+        )
+
+        nevents = len(events)
+        sampling_ratio_avg = sum(event.capture_pct for event in events) / nevents / 100.0
```
Comment (majorgreys) [inline ddtrace/profiling/exporter/pprof.py:149]: We could shave off some effort by collapsing these three iterations over events into one.

**Review by majorgreys — COMMENTED** [review]
Just some minor questions

File: ddtrace/profiling/exporter/pprof.py (hunk 2 above)
Comment (jd) [reply ddtrace/profiling/exporter/pprof.py:149]: You'd think so…

```python
l = list(range(10_000_000))

def v1():
    return sum(l), sum(l), sum(l)

def v2():
    x = y = z = 0
    for v in l:
         x += v
         y += v
         z += v
    return x, y, z


assert v1() == v2()


import timeit

print(timeit.timeit("v1()", number=10, globals=globals()))
print(timeit.timeit("v2()", number=10, globals=globals()))
```

v2 is 6 times slower:
```
$ python3 t.py
2.568303665
14.879553143
```

So I'm not sure it'd be a win even here 🤔

File: ddtrace/profiling/collector/memalloc.py (hunk 1 above)
Comment (jd) [reply ddtrace/profiling/collector/memalloc.py:64]: They're smaller and we don't need to modify the result, therefore tuple is good enough :)

File: ddtrace/profiling/exporter/pprof.py (hunk 2 above)
Comment (majorgreys) [reply ddtrace/profiling/exporter/pprof.py:149]: Should have actually tested out my suggestion before making it! 🔬 

Funny but my initial thought was to use `functools.reduce` but that's even slower:

```
def v3():
    return functools.reduce(lambda x,y: (x[0]+y, x[1]+y, x[2]+y), l, (0,0,0))
```

Which takes ~21s.

**Review by majorgreys — APPROVED** [review]

**Pull request merged**